File: include/apl/types/Settable.java
package APL.types;

import APL.Type;
import APL.errors.ValueError;

public abstract class Settable extends Obj {
  final Obj v;
  protected Settable(Obj v) {
    this.v = v;
  }
  @Override
  public Type type() {
    return v == null? Type.var : v.type();
  }
  public abstract void set(Obj v, Callable blame);
  public Obj get() {
    if (v == null) throw new ValueError("trying to get value of non-existing settable", this);
    return v;
  }
  public Obj getOrThis() {
    if (v == null) return this;
    return v;
  }
}

File: APL/types/Settable.cpp
#include <APL/Settable.h>

namespace APL::types
{
	using Type = APL::Type;
	using ValueError = APL::errors::ValueError;

	Settable::Settable(std::shared_ptr<Obj> v) : v(v)
	{
	}

	APL::Type Settable::type()
	{
	  return v == nullptr? Type::var : v->type();
	}

	std::shared_ptr<Obj> Settable::get()
	{
	  if (v == nullptr)
	  {
		  throw ValueError(L"trying to get value of non-existing settable", shared_from_this());
	  }
	  return v;
	}

	std::shared_ptr<Obj> Settable::getOrThis()
	{
	  if (v == nullptr)
	  {
		  return shared_from_this();
	  }
	  return v;
	}
}
